### DISCUSSION TITLE
Troubleshooting developer install

### DISCUSSION THREAD

**Discussion by darryl-ad** [pybamm-team/PyBaMM#2988] [General]
Hello! I had some trouble installing PyBaMM from source with the developer options [following these instructions](https://pybamm.readthedocs.io/en/latest/source/user_guide/installation/install-from-source.html), so I thought I'd share the solutions I found along the way :) 

### Windows Subsytem for Linux (WSL)

I initially wanted to install in Windows, but it was too hard to follow, so I then reverted to WSL. The first hurdle was that the most common option, Ubuntu, ships with Python 3.10 as default, whereas PyBaMM supports 3.7-3.9. I realised this a while later, and installed 3.9 and selected it as my default. However it still caused too many issues (in particular getting `distutils` to work), so I gave up on that. PS - **never** uninstall the default version of Python in Ubuntu 😂

Next step was to use Debian instead of Ubuntu. This allowed me to get a clean Python 3.9 install, but was significantly harder to do so. I have lost much of the troubleshooting I did around this, but it mostly involved getting Debian to recognise the version of Python installed!

### Install Instructions

The instructions specify to install the following set of pre-requisites:

> sudo apt install python3.X python3.X-dev libopenblas-dev gcc gfortran

But I found I also needed the following in order to pass the unit tests:

> pip3 install lcapy

> sudo apt install ghostscript texlive-latex-extra cm-super

> sudo apt-get install dvipng

> sudo apt-get install graphviz

`graphiz` was the hardest to find as it produced a generic error in the unit test, but I ended up finding it [here](https://github.com/pybamm-team/PyBaMM/blob/develop/pybamm/expression_tree/symbol.py#LL475C59-L475C66).

Finally, after creating and activating the virtual environment with Tox, I found that I then had to install Tox for a 2nd time to run the tests:

> python3.9 -m pip install "tox<4"

And I think that about fixed it!
Top recommendation is just to use a Mac lol.

**Comment by agriyakhetarpal** [comment]
Hi there, the current version of PyBaMM supports all Python versions from 3.8 to 3.11 on all platforms, so installing with Python 3.10 should work out of the box. 

To get the unit tests that fail due to LaTeX-related errors to pass, it is recommended to install the `texlive-full` package.

When you create a virtual environment, it is isolated from the global Python installation and therefore does not come with any dependencies inside it, so `Tox` is not included by default—it needs to be installed in the virtual environment with `pip3.X install "tox<4"`.

**Reply by darryl-ad** [reply]
Hello, yeah you're right I was looking at an out-of-date source for the Python version support my bad!

It would be useful to add the `texlive-full` package to the pre-requisites [here](https://pybamm.readthedocs.io/en/latest/source/user_guide/installation/install-from-source.html), along with `dvipng` and `graphviz` which I also needed to get manually.

I understand installing packages within virtual environments, but I mean that when you create the `Tox` dev virtual environment `tox -e dev `, the only package that needs reinstalled at that point is `Tox` itself...which is a bit counterintuitive!

> This creates a virtual environment .tox/dev (or windows-dev) inside the PyBaMM/ directory. It comes ready with PyBaMM and some useful development tools like [pre-commit](https://pre-commit.com/) and [black](https://black.readthedocs.io/en/stable/).

Maybe it would be possible to add `Tox` to the list of useful development tools mentioned above^.

**Reply by agriyakhetarpal** [reply]
Yes, I agree that the installation instructions are a bit confusing. There are some discrepancies between instructions for macOS and Linux too—for example, the [Developer install](https://pybamm.readthedocs.io/en/latest/source/user_guide/installation/GNU-linux.html#developer-install) section correctly lists the usage of the `apt install graphviz` command for Linux, but there is no mention of it in the instructions for installing from source (which are platform-agnostic). Similarly, the installation of `texlive-full` is not mentioned in the documentation, but it does get installed in the workflows that run on pull requests. 

I am not sure whether `tox` should be included inside the environment as a recursive dependency, since I usually run the test suite with `tox -e tests`, and this is when I am working outside the `dev` virtual environment and I have a user installation of `tox` outside both `dev` and `tests` which is global to the system (as with `pip3.X install --user "tox<4"` it installs packages and libraries to the home directory of the current user).
